### PR TITLE
Another pass at basic content models

### PR DESCRIPTION
**Pull request by ormsbee**
This splits Item-related models into a new app and otherwise follows the logical split discussed [here](https://github.com/openedx/openedx-learning/issues/1#issuecomment-1134873686).

### CODE REVIEW

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:40]: I tried to give the rationale here: https://github.com/openedx/openedx-learning/pull/13/files#diff-7844f3ef353b5f0e84d0cfdcd0798d9a0922af8d369dfabd7ce35636d9e4b22eR80-R96

Comment (feanil) [reply openedx_learning/apps/core/itemstore/models.py:40]: Makes sense thanks!

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/apps/core/itemstore/models.py:65]: Why is `title` here and not in ItemInfo? Is it because it's example of policy, which we're distinguishing from metadata (like `type`)?

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/apps/core/itemstore/models.py:36]: knowledge check: this is secondary source of information, since one could authoritatively determine this by asking the `publish` app what the published LearningContextVersion is, and then looking at the ItemVersions within that LCV, right?

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/apps/core/itemstore/models.py:10]: Ack, sorry, I am struggling to rebuild this hierarchy in my head. I'm going to try to rephrase it here. Let me know if I'm off:
* ...Assume "LC" is some LearningContext, and "LCV" some LearningContextVersion within LC.
* An ItemRaw is a blob of byte data and a mimetype.
  * Example: an ItemRaw instance "PR", containing some OLX file
* An ItemInfo associates an ItemRaw with metadata, including its LearningContext and its item type.
  * Example: an ItemInfo instance "PI", which tells us that PR is a `problem` and belongs to LC.
* An Item is an atomic block of content associted with a LC. In isolation, an Item does not tell us about its content, because content is associated with versions and an Item is versionless.
  * Example: an Item instance "P" in LC.
 * An ItemVersion joins together an abstract Item instance with a concrete ItemInfo, denoting a particular version of the content, policy, and metadata of an Item.
   * Example: an ItemVersion instance "PV", joining together P and PI.
   * Example: an ItemVersion instance "PV2", joining together P and PI, with some updated policy.
   * Example: a later ItemVersion instance "PVx", joining together P and some modified ItemInfo "PIx".
   * and so on
 * LearningContextVersionItemVersions relate LearningContextVersions to ItemVersions. This table is the connective tissue telling us which ItemVersions are in a LearningContextVersion (and, by corollary, which Items are in a LearningContext).
   * Example: PV is in LCV
   * Example: PV2 is in "LCV2", a later version of LC
   * Example: PV2 is also in "LCV3", a later version of LC that updates some Item(s) other than P

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/apps/core/itemstore/models.py:85]: Can we always deduce `type` from `item_raw.mime_type`?

**Review by kdmccormick — COMMENTED** [review]
Sorry I took so long to review this. Most of my comments are just me trying to make sure I grok everything.

File: projects/dev.py
Comment (ormsbee) [reply projects/dev.py:37]: > The `apps` namespace package seems unnecessary and out-of-step with other edx-platform libraries. Why not just `openedx_learning.core.itemstore`?

Mostly because I wanted a place to hang `lib` for things like the field types.

> I notice that you are using verbs for some of the core app names. Wouldn't it be more standard to use nouns? (`composition`, `publishing`)

Yeah, I was going for the verb form for a while before because it was shorter, before I landed on `itemstore` for one of them. I'll switch it back over to nouns.

File: olx_importer/management/commands/load_course_data.py
Comment (ormsbee) [reply olx_importer/management/commands/load_course_data.py:171]: Ah, right. I guess it doesn't hurt to add the v1, though I'm skeptical we'll ever have a v2. (I'm guessing we'd cali it something entirely different, to avoid confusion.) Will fix.

File: olx_importer/management/commands/load_course_data.py
Comment (ormsbee) [reply olx_importer/management/commands/load_course_data.py:198]: Ugh, right, because I didn't ever put the static asset mapping stuff back. Thank you.

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:46]: Yeah. I think that in general, the `identifier` fields have to be namespaced so that it's entirely local to one particular LearningContext.

Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:46]: In the case of XBlocks, do you see this `identifier` as being the usage key (`block-v1:a+b+c+type@video+block@xyz`) or the block id (just `xyz`)?

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:28]: > Is this the LearningContext that the Item is being used in, or the LearningContext that it's defined in (eg, a content library)? Is that distinction something that even exists at this level?

The one it's being used in, because Item has the identifier, and we need to be able to hang course-level policy changes off of ItemVersion. So the hierarchy kinda looks kind of like this:

```mermaid
graph TD;
    CI[Course Item]-->CIV[Course ItemVersion];
    CIV-->LII[Library ItemInfo];
    LII-->LIR[Library ItemRaw];
    LI[Library Item]-->LIV[Library ItemVersion];
    LIV-->LII;
```

So the course makes a copy of the high level objects (`Item`, `ItemVersion`), and shares the lower-level ones (`ItemInfo`, `ItemRaw`, and anything derived from those).

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:36]: Yes. And maybe it's not a great idea to have redundancy like this. The main motivation for this was that a lot of the time, you want to bind data against the abstract notion of a piece of content across all its versions (e.g. most student state), and I wanted a really convenient way to get to the current "live" version from there.

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:65]: Yes. Renaming is desirable for libraries, because you don't know what kind of convention that library is using, e.g. "Problem 22a" may not make as much sense for the new context it's being used in.

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:85]: I had originally meant it as something different, but I'm not sure if that holds up in the latest iteration.

The difference I had in mind originally was something like an HTMLBlock's HTML. That it would have an `ItemInfo` that showed it was an XBlock with HTML handler, but the raw data had a MIME type of "text/html".

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:110]: Yes. It's mostly there for ownership/cleanup.

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:28]: That makes sense! The graphic is immensely helpful. I have to admit that the names of the models on their own did not guide me to this understanding.

The split between [Item , ItemVersion] and [ItemInfo, ItemRaw] reminds me of the XBlock framework's distinction between Usage and Definition, respectively. Is that a fair parallel?

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:110]: Cool. This would be helpful to note in a comment.

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:65]: Makes sense!

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:36]: That does sound convenient. I guess, you're either going to end up needing to keep this field up-to-date, or you're going to end up with a lot of requests making the same query ("what's the published version?"). I don't think I have the experience to know which of these is an easier challenge to manage.

File: projects/dev.py
Comment (kdmccormick) [reply projects/dev.py:37]: > Mostly because I wanted a place to hang lib for things like the field types.

The 2nd-level packages could be `openedx_learning.core`, `openedx_learning.lib`, and `openedx_learning.contrib`.

I remember as a new dev being frazzled by the long import paths of `openedx.core.djangoapps...` in edx-platform. I could never remember the paths; I was always copying and pasting, and it made the platform feel more overwhelming.

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/apps/core/itemstore/models.py:103]: @ormsbee I am struggling a bit with the distinction between ItemInfo and ItemRaw.

What would we lose by collapsing them into a single model? Is there something bad about structured metadata models hanging off of the same model that's storing the raw bytes?

Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:103]: I think I originally created the idea of this model (under a different name), when I was thinking it would have more stuff in it (like a local identifier associated with the content package), but those concepts were simplified out of the design. We would have to be careful to not pull back the raw payload all the time, but that's manageable.

The only remaining use case I can think of is when you want to add a little bit of metadata about the content that is not the content itself–new information entered by a human, not something that's being derived by the code from an `ItemRaw`. At that point, we have to figure out where to hang that data in a way that also clearly signals that this content has changed.

As an example, let's say we have a text description of an image.

*If we attach that data to a model that hangs off of `ItemRaw`*, then there's no indication that the content has been updated. None of the versioning mechanisms above this were invoked or needed to be invoked. So we've done an unexpected mutation, and it will invalidate assumptions around versioning and caching. We can't even create a new `ItemRaw` to bind it to (wasteful as that would be in terms of space), because `ItemRaw` uses a hash of its contents for identity.

*We could attach that data to a model that hangs off of `ItemVersion`.*  That's a little weird because that's not policy data that changes from course to course, but we would have default policy information that's attached at that layer and copied, so maybe it would work out...?

You're right in that other comment, it's very much like the definition/usage split in the old system. But maybe we should take that as a cautionary tale–people put many things in the settings scope that should have been put in definitions. If the distinction of when to use `ItemInfo` is unclear to you as an expert in this kind of system, then it will almost certainly confuse others who try to use this later. Maybe it's easier to say: "If you can completely derive it from the raw data, hang a model off of `ItemRaw`, and if it's any new data entered in by a user, hang it off of `ItemVersion`."?

File: projects/dev.py
Comment (ormsbee) [reply projects/dev.py:37]: Okay, I'll give that a shot. I do like not having `apps` appear twice in that path (package + apps.py). Thank you!

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py:46]: I'd prefer not to have the full learning context key encoded, but since we currently allow blocks of different types to have the same ID, I think we have to encode at least something like `video+xyz`

**Comment by ormsbee** [comment]
FYI @bradenmacdonald: This is the data models PR I briefly mentioned in our meeting today.

File: openedx_learning/apps/core/itemstore/models.py
Comment (kdmccormick) [reply openedx_learning/apps/core/itemstore/models.py:103]: I just realized I need to check an assumption about how I've been reading this: "hanging model X off of model Y" means that X would have a foreign key to Y, right?

File: openedx_learning/apps/core/itemstore/models.py
Comment (bradenmacdonald) [reply openedx_learning/apps/core/itemstore/models.py:36]: What about making `published_version` a [memoized] python getter that uses the publish app to return the published version? Or do you mean specifically you want to be able to use the django ORM methods on this `published_version` field so you can `select_related` and do things like filtering queries?

File: openedx_learning/apps/core/itemstore/models.py
Comment (bradenmacdonald) [reply openedx_learning/apps/core/itemstore/models.py:65]: Should this be `title_override` then to clarify that you can optionally stick with whatever the library author gave as a title?

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py]: > Or do you mean specifically you want to be able to use the django ORM methods on this published_version field so you can select_related and do things like filtering queries?

Yeah, exactly this. I want to make sure we can still do efficient filtering queries. But I think I'll actually leave this out for now and add it in later if it becomes necessary. Even if we do keep this kind of a key, it might make more sense in a separate model that specifically separates the publishing details (`published_version`, `first_published_at`, `last_published_at`, etc.).

File: openedx_learning/apps/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/apps/core/itemstore/models.py]: I'm inclined to not use the `_override` convention because the default content in the library would also be using this field (just under the content library's learning context).

Though maybe the title should be lifted out into a separate model? Maybe even one with language awareness?

Wait... maybe ItemRaw:ItemVersion is M:1?

I'll make comments along both of those lines for future followup, but won't make any immediate model changes.

**Comment by kdmccormick** [comment]
Hey @ormsbee , are you looking for another round of review on this one?

**Comment by ormsbee** [comment]
@kdmccormick: Almost. I've been doing bits and pieces in the odd-hours for this. I need to update the import script to make sure these ideas all hold together and then it'll be ready for review. Though I don't expect the models.py files to change very much at this point, so if you want to take another pass at that, you're more than welcome to.

**Comment by ormsbee** [comment]
@kdmccormick: ready for another round of review, with the caveat that I'd ask that you largely ignore `load_course_data.py`  and the `staticassets` app at this point. I wanted to make sure it was functional (i.e. the data model wasn't totally broken), but the interface is completely wrong for the import script and `staticassets` is pretty half-baked at the moment.

File: openedx_learning/core/itemstore/models.py
Comment (kdmccormick) [inline openedx_learning/core/itemstore/models.py:176]: > maybe we should split it up into two fields instead of doing substring searches? 

Sounds like a win to me. Any downsides?

Another reason to split: [MIME types can include a `parameter` field](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#structure_of_a_mime_type), eg `text/plain;charset=utf8`. If you don't want to support the parameter field, then splitting into two fields would make that more explicit.

File: openedx_learning/core/itemstore/models_api.py
Comment (kdmccormick) [inline openedx_learning/core/itemstore/models_api.py:11]: So, I'm reading that:
* each ItemVersion can only be associated with at most one of each type of "Data" model (Image, Asset, Video, etc.), but
* each ItemVersion may be associated with an arbitrary number of Content objects.

This confused me at first, because the "Data" models and the Content model occupy very similar roles in my head. But I think I'm beginning to understand it... is this a good example?
* Some ItemVersion represents a video.
  * It has multiple Content items associated with: one that points to the video file, one PNG thumbnail, and several others text files that are transcripts.
  * In the staticassets app, it has a single associated Video instance and Asset instance. Video metadata as well as information about the transcripts are part of (or hang off) of the Video model.

File: openedx_learning/core/itemstore/models_api.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models_api.py:12]: > is Data immutable

It should be or Bad Things will happen.

> is ItemVersionData:Data an M:1 relationship because multiple versions of an Item may reference the same underlying Data?

Yes. One of the underlying assumptions here is that if you have some data that's associated with an ItemVersion, changing that data means that the Item has changed in some way, and thus forms a new ItemVersion. So if you change the retry-settings or some such for an Item, there's a new ItemVersion. But for all the other apps that didn't change their data, we can just make new entries that join the new ItemVersion with the same data that those apps had for the last ItemVersion.

File: openedx_learning/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models.py:176]: I guess I'm just used to thinking them as a combined thing because I can't really see querying one in isolation from the other. But it is more normalized and it's easy to add, so 🤷 I'll add it.

File: openedx_learning/core/itemstore/models_api.py
Comment (kdmccormick) [reply openedx_learning/core/itemstore/models_api.py:12]: @ormsbee 

> It should be or Bad Things will happen.

Cool, I assumed so but wanted to be sure. Is there a way to indicate that, like marking it read-only in Django admin?

> Yes. ...

Makes sense 👍🏻 It would be good to include that explanation somewhere in the library before releasing v1.0. Not a blocker for this PR; I know you're still iterating.

File: openedx_learning/core/itemstore/models.py
Comment (feanil) [inline openedx_learning/core/itemstore/models.py:38]: So an item can only be in one `LearningContext`(LC), but since an `ItemVersion` can be in any number of `LearningContextVersions`(LCV) do we need a constraint somewhere that ensures that all those LCVs are from the same LC?  Or is that too paranoid at this stage?

File: openedx_learning/core/itemstore/models.py
Comment (feanil) [inline openedx_learning/core/itemstore/models.py:175]: When would we need to do a partial search? Is it just when looking at types and ignoring sub-types? or something more complex?

**Review by feanil — APPROVED** [review]
A couple of questions but nothing blocking this from landing.  The re-organization generally makes sense to me based on previous conversations.

File: openedx_learning/core/itemstore/models_api.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models_api.py:11]: Yeah, that was my intent. A couple of extra notes about it:

### Content survives after apps die.

At some point, an app becomes unsupported and disappears. At that point, all the Item/ItemVersion-related data it has dies as well. But Content is associated with the ItemVersion directly in `itemstore`, so that raw data will still export.

So I think this encourages a pattern where we store raw things in Content and then build smarter relational data structures out of them connected to ItemVersion. For instance, a ProblemBlock's raw OLX can live in Content while richer relational tables live in a new ProblemBlock app that give a richer, more queryable interface.

### ItemVersion mapping to just one row in a "Data" model is common, but not required.

I think that most of the time apps want to associate one piece of data with a given ItemVersion (e.g. a problem weight, schedule, etc.), but that's not a requirement. A rich data model of a ProblemBlock might want to create entries for each response, for instance.

File: openedx_learning/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models.py:70]: Yeah, that was my thought as well.

File: openedx_learning/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models.py:78]: Yeah, I think you're right that it belongs in a higher level. In any case, I'm definitely fine with punting this for now, even if we do bake it at this layer later. I'll remove the comment.

File: openedx_learning/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models.py:175]: Yeah, that's the only situation I can really think of.

File: openedx_learning/core/itemstore/models.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models.py:38]: I like being paranoid about the data model in general, but I'm not sure what we can do with MySQL here, or how much overhead adding CheckConstraints is going to cause. I'll see how hard it is.

File: openedx_learning/core/itemstore/models_api.py
Comment (kdmccormick) [reply openedx_learning/core/itemstore/models_api.py:11]: > ItemVersion mapping to just one row in a "Data" model is common, but not required.

How would that look?

For example, imagine we have an ItemVersion representing a problem block, and Response data models that we want to (indirectly) hang off of it. Would it look like:

```
ItemVersion  <--(1:1)-- ItemVersionProblemBlock --(N:N)--> Response
```
or:
```
ItemVersion  <--(1:N)-- ItemVersionResponse --(N:1)--> Response
```
or something else?

Comment (ormsbee) [reply openedx_learning/core/itemstore/models_api.py:11]: I guess I thought it would be like:

```
ItemVersion <-- (N:1) -- ItemVersionProblemBlock -- (N:N) --> Response
```

That would still allow some central process to increment to the next version in a simple way without understanding what ItemVersionProblemBlock is. But I honestly haven't thought it through that much.

File: openedx_learning/core/itemstore/models_api.py
Comment (ormsbee) [reply openedx_learning/core/itemstore/models_api.py:12]: > Cool, I assumed so but wanted to be sure. Is there a way to indicate that, like marking it read-only in Django admin?

Yeah, I'll definitely do that when I start making some admin pages.

**Pull request merged**
